### ISSUE TITLE
Range Queries

### ISSUE THREAD

**Issue by Dan-Knox** [bluejekyll/foundationdb-rs#115]
I apologize in advance if this isn't the appropriate place for this question but I was hoping you could provide me with a few example range queries. I am brand new to Rust and can't seem to figure out the proper API for range queries. Here is what I have tried so far...

```rust
    let start = foundationdb::keyselector::KeySelector::new(b"hello".to_vec(), true, 0);
    let end = foundationdb::keyselector::KeySelector::new(b"hello".to_vec(), true, 0);
    let range_options = foundationdb::transaction::RangeOptionBuilder::new(start, end).limit(10).build();

    trx = fdb.db.create_trx().expect("failed to create transaction");
    let range = trx.get_range(range_options, 10);
    let range_result = range.wait().expect("something horribad");
    let keyvalues: foundationdb::future::KeyValues = range_result.key_values();

    match keyvalues.first() {
        Some(value) => println!("key: {}", String::from_utf8_lossy(value.key())),
        None => println!("none :(")
    }
```

I know the key `hello` exists since I pretty much copied and pasted the example from the README of the documentation which specifically inserts and retrieves that key. That code of course works great for me. My range query however returns zero results no matter what I do.

Any help would be greatly appreciated.

**Comment by bluejekyll** [comment]
From the docs, I think your end KeySelector might be incorrect:

> ... which have a key lexicographically greater than or equal to the key resolved by the begin key selector and lexicographically less than the key resolved by the end key selector.

I think if you try end as "hello world" or "hellp", you might get different results? i.e. the query is exclusive, not inclusive: `[start, end)`

**Comment by Dan-Knox** [comment]
Alas I did try the "hellp" as the ending key and still no dice. It looked to you like I was using it correctly though?

**Comment by bluejekyll** [comment]
I don’t see anything obviously wrong... here’s the test case, which might be helpful: https://github.com/bluejekyll/foundationdb-rs/blob/master/foundationdb/tests/range.rs

**Comment by Dan-Knox** [comment]
I restructured my code to match the test case and everything is working now. I was constructing the key selectors directly with the `::new` method instead of using the `::first_or_*` methods which does not return any results.

One last question, is it possible to use the `Tuple` struct to pack keys in parity with the other FoundationDB drivers such as the ruby or python driver? I don't see any test cases for the `Tuple` structs. I am hoping I will be able to pack a tuple from ruby and then retrieve it from the rust code.

Thank you for your time on this. You have already been incredibly responsive and helpful!

**Comment by bluejekyll** [comment]
> Using `::new` directly does not return any results.

Hm. that's either a bug, or something we should probably remove.

> is it possible to use the `Tuple` struct to pack keys in parity with the other FoundationDB drivers such as the ruby or python driver?

We worked pretty hard to make the Tuple types consistent with the FDB definitions, so as long as Ruby and Python are correct then it should be good. If I remember correctly, Python strings may have some issues, which I raised in a thread a while ago on the FDB forums, where I think they may not be fully portable:

https://forums.foundationdb.org/t/rust-client-api-0-2-0/386/2?u=bluejekyll
